server/odc-service/src/main/java/com/oceanbase/odc/service/iam/UserService.java

```diff
@@ -192,44 +190,55 @@ public class UserService {
-        UserEntity userEntityCreated = userRepository.findByAccountName(userEntity.getAccountName()).get();
-        Long userId = userEntityCreated.getId();
-        User user = new User(userEntityCreated);
-        if (Objects.nonNull(roleNames)) {
-            List<Long> roleIds = new ArrayList<>();
-            List<Role> roles = new ArrayList<>();
-            for (String roleName : roleNames) {
-                Optional<RoleEntity> roleEntity = roleRepository.findByNameAndOrganizationId(roleName, organizationId);
-                PreConditions.validExists(ResourceType.ODC_ROLE, "roleName", roleName, roleEntity::isPresent);
-                PreConditions.validArgumentState(roleEntity.get().getType() != RoleType.INTERNAL,
-                        ErrorCodes.BadArgument,
-                        new Object[] {"Internal role does not allow operation"},
-                        "Internal role does not allow operation");
-                Long roleId = roleEntity.get().getId();
-                UserRoleEntity userRoleEntity = new UserRoleEntity();
-                userRoleEntity.setUserId(userId);
-                userRoleEntity.setRoleId(roleId);
-                userRoleEntity.setOrganizationId(organizationId);
-                userRoleEntity.setCreatorId(OdcConstants.DEFAULT_ADMIN_USER_ID);
-                userRoleRepository.saveAndFlush(userRoleEntity);
-
-                roleIds.add(roleId);
-                roles.add(new Role(roleEntity.get()));
+    public User createIfNotExistsOrUpdate(UserEntity userEntity, List<String> roleNames) {
+        User user;
+        Optional<UserEntity> optionalUser = userRepository.findByAccountName(userEntity.getAccountName());
+        if (optionalUser.isPresent()) {
+            User existed = new User(optionalUser.get());
+            if (!Objects.equals(existed.getExtraProperties(), userEntity.getExtraPropertiesJson())) {
+                user = save(userEntity);
+                log.info("Update user extra properties successfully, accountName={}, new properties={}",
+                        user.getAccountName(), user.getExtraProperties());
+            } else {
+                user = existed;
+                log.info("User already exists, accountName={}", existed.getAccountName());
+            }
+        } else {
+            user = save(userEntity);
+            if (CollectionUtils.isNotEmpty(roleNames)) {
+                List<Long> roleIds = new ArrayList<>();
+                List<Role> roles = new ArrayList<>();
+                for (String roleName : roleNames) {
+                    RoleEntity roleEntity =
+                            roleRepository.findByNameAndOrganizationId(roleName, user.getOrganizationId())
+                                    .orElseThrow(() -> new NotFoundException(ResourceType.ODC_ROLE, "name", roleName));
+                    PreConditions.validArgumentState(roleEntity.getType() != RoleType.INTERNAL, ErrorCodes.BadArgument,
```

Comment:
Prefer use `Verify` instead of `PreConditions`

Reply:
It is no matter, `PreConditions` also throw exception with i18n message.

Reply:
It's not about the effect, it's about the usage.
We generally use `PreConditions` at the very beginning of a method for parameter validation, while `Verify` is typically used for logical validation.

Reply:
Okay, done!